llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp

```diff
@@ -1571,7 +1583,11 @@ simplifySVEIntrinsicBinOp(InstCombiner &IC, IntrinsicInst &II,
   else
     SimpleII = simplifyBinOp(Opc, Op1, Op2, DL);
 
-  if (!SimpleII)
+  // An SVE intrinsic's result is always defined. However, this is not the case
+  // for its equivalent IR instruction (e.g. when shifting by an amount more
+  // than the data's bitwidth). Simplifications to an undefined result must be
+  // ignored to preserve the intrinsic's expected behaviour.
+  if (!SimpleII || isa<UndefValue>(SimpleII))
```

Comment:
There are multiple ways that `undef` (or `poison`?) can be returned, for example `Op1` itself could be undef. Do we have a test for this case? I assume you're taking the conservative approach here and just ignoring safe optimisations like `asr <vscale x 4 x i32> undef, <vscale x 4 x i32> splat(i32 0)` for now to reduce code complexity?

Also, is it always guaranteed to be `UndefValue` - do we ever return poison?

Reply:
I talk to this in the commit message. At this stage we've not encountered any legitimate reasons to propagate poison so they are ignored to reduce complexity.  As there was no intent to propagate poison from my original code there are no tests that rely on it. Which is the other reason it seemed reasonable to make the switch.

I believe `isa<UndefValue>(X)` returns true for both `undef` and `poison` because they are both types of undefined values.

Reply:
I've been thinking about this some more and I'm now leaning towards saying the SVE intrinsics should not propagate poison in general.  Whilst the IR is fully define in this respect the SVE ACLE is not and I think some of the IR's outcomes would not match ACLE writer expectations.

Reply:
I've been cleaning up the SVE intrinsics combine tests so at the same time I also added a test for poison propagation. I've rebased this PR so that it shows the new expected behaviour (i.e. we shouldn't propagation poison).